File: python/chain_of_responsibility/chain_of_responsibility.py
from abc import ABC


class SumHandler(ABC):
    def __init__(self, next_handler=None):
        self._next_handler: SumHandler = next_handler

    def sum(self, a: int, b: int) -> int:
        raise NotImplementedError()


class PositiveSumHandler(SumHandler):
    def sum(self, a: int, b: int) -> int:
        if a > 0 and b > 0:
            return a + b
        if self._next_handler is not None:
            return self._next_handler.sum(a, b)
        else:
            return 0


class StrangeSumHanfler(SumHandler):

    def sum(self, a: int, b: int) -> int:
        return 42


def main():
    handler3 = StrangeSumHanfler()
    handler2 = PositiveSumHandler(handler3)
    handler1 = PositiveSumHandler(handler2)
    print(handler1.sum(-1, 3))
    print(handler1.sum(3, 3))


if __name__ == '__main__':
    main()

File: cpp/chain_of_responsibility/chain_of_responsibility.cpp
#include <iostream>
#include <memory>
#include <utility>

class SumHandler {
public:
    virtual int sum(int a, int b) = 0;

    void set_next_handler(std::shared_ptr<SumHandler> handler) {
        _next_handler = std::move(handler);
    }

protected:
    std::shared_ptr<SumHandler> _next_handler;
};

class PositiveSumHandler : public SumHandler {
public:
    int sum(int a, int b) override {
        if (a > 0 && b > 0) {
            return a + b;
        }
        if (_next_handler != nullptr) {
            return _next_handler->sum(a, b);
        }
        return 0;
    }
};

class StrangeSumHandler : public SumHandler {
public:
    int sum(int a, int b) override {
        return 42;
    }
};

int main() {
    auto handler1 = std::make_shared<PositiveSumHandler>();
    auto handler2 = std::make_shared<PositiveSumHandler>();
    auto handler3 = std::make_shared<StrangeSumHandler>();
    handler1->set_next_handler(handler2);
    handler2->set_next_handler(handler3);
    std::cout << handler1->sum(-1, 2) << std::endl;
    std::cout << handler1->sum(3, 2) << std::endl;
    return 0;
}
